src/python/pants/rules/core/list_target_types.py

```diff
@@ -59,7 +72,35 @@ class FieldInfo:
 
     @classmethod
     def create(cls, field: Type[Field]) -> "FieldInfo":
-        description = get_docstring(field, flatten=True)
+        # NB: It is very common (and encouraged) to subclass Fields to give custom behavior, e.g.
+        # `PythonSources` subclassing `Sources`. Here, we set `fallback_to_ancestors=True` so that
+        # we can still generate meaningful documentation for all these custom fields without
+        # requiring the Field author to rewrite the docstring.
+        #
+        # However, if the original `Field` author did not define docstring, then this means we
+        # would typically fall back to the docstring for `AsyncField`, `PrimitiveField`, or a
+        # helper class like `StringField`. This is a quirk of this heuristic and it's not
+        # intentional. So, we filter out the docstring for these abstract classes.
```

Comment:
Can you use `inspect.isabstract` here more generally? That said, why not just take any doc string you can get? If the fallback is to a user supplied TimeoutField which is abstract (say they designed it to require subclassing to specify a max timeout) and the subclass does this but does not define a docstring - it seems we'd want to take the TimeoutField docstring as better than further backing up to FloatField or None.

Reply:
I tried using `inspect.isabstract`, which failed because it needs an object and not a class.

Then I used `hasattr(cls, "__abstractmethods__")` but it didn't work properly because concrete subclasses still had the dunder.

--

> If the fallback is to a user supplied TimeoutField which is abstract (say they designed it to require subclassing to specify a max timeout) and the subclass does this but does not define a docstring - it seems we'd want to take the TimeoutField docstring as better than further backing up to FloatField or None.

Agreed, which is precisely what we are able to do now.

The other important edge case is that if both `MyTimeoutField` and `TimeoutField` leave off docstring, we should not fall back to `FloatField` or `PrimitiveField`.

Reply:
Hmm:
```python
$ python3.6
Python 3.6.10 (default, Jan 21 2020, 16:15:00) 
[GCC 9.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import inspect
>>> from abc import ABC, abstractmethod
>>> class Field(ABC):
...   @abstractmethod
...   def name(self) -> str:
...     pass
... 
>>> class Bob(Field):
...   pass
... 
>>> class Billy(Bob):
...   def name(self) -> str:
...     return "Billy"
... 
>>> inspect.isabstract(Field)
True
>>> inspect.isabstract(Bob)
True
>>> inspect.isabstract(Billy)
False
>>> 
```

Why wouldn't you want to fall back to FloatField docs if it had them?

Reply:
TIL that subclassing `ABC` isn't enough - you must actually define `@abstractmethods` to make it abstract.

```python
▶ python3.6
Python 3.6.10 (default, Feb 26 2020, 08:26:13)
[GCC 4.2.1 Compatible Apple LLVM 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import abc
>>> import inspect
>>> class A1(abc.ABC):
...   pass
...
>>> inspect.isabstract(A1)
False
>>> A1()
<__main__.A1 object at 0x107c75198>
```

--

> Why wouldn't you want to fall back to FloatField docs if it had them?

Because the docstring for `PrimitiveField`, `FloatField`, etc. is oriented towards field authors, rather than end-users.

For example, if I remove the docstring from `Compatbility` and remove this filtering, we would get this confusing and leaky generated documentation:

![Screen Shot 2020-03-30 at 10 20 06 AM](https://user-images.githubusercontent.com/14852634/77942105-09877f80-7270-11ea-923b-d358fa42d7f7.png)

Reply:
Aha! That is a much better answer. I'll send up a PR to correct the comment.

Reply:
Ah, I see now why the comment obfuscated that. Thank you!